### PR TITLE
[Enabler][zos_data_set] Add support to GDG and special characters.

### LINKED ISSUE #1371: [Enabler] [zos_data_set] GDG/GDS and special character support
# Enabler description

New MVSDataset class
-----------------------

There is a new MVSDataset class introduced in #1386 that represents a data set, when used, as part of initialization it escapes and resolves the gds relative name if any.

This is a sample on how to initialize the new class, many of these params can just be None.

```python
MVSDataSet(
            name = params.get("name"),
            record_format = params.get("record_format", None),
            volumes = params.get("volumes", None),
            data_set_type = params.get("type", None),
            block_size = params.get("block_size", None),
            record_length = params.get("record_length", None),
            space_primary = params.get("space_primary", None),
            space_secondary = params.get("space_secondary", None),
            space_type = params.get("space_type", None),
            directory_blocks = params.get("directory_blocks", None),
            key_length = params.get("key_length", None),
            key_offset = params.get("key_offset", None),
            sms_storage_class = params.get("sms_storage_class", None),
            sms_data_class = params.get("sms_data_class", None),
            sms_management_class = params.get("sms_management_class", None),
        )
```

GDG/GDS Support
-------------------

Option `name` accept a data set name which depending on the `state` option is whatever depends 

Internally, in the case of a data set name being a GDS relative name we should resolve those names into absolute GDS names when the generation number is 0 or negative when `state=catalog` or `state=absent`.

Following functions will be implemented in #1386 
```python
# Sample of soon to be implemented gdg function
if DataSet.is_gdg():
    src = DataSet.resolve_gdg_absolute_name(src)
```

```yaml
- name: Create a sequential data set if it does not exist
  zos_data_set:
    name: ZOAUSER.GDG(+1)
    type: SEQ
    state: present

- name: Create a sequential data set if it does not exist
  zos_data_set:
    name: ZOAUSER.GDG(-3)
    state: absent

- name: Create a sequential data set if it does not exist
  zos_data_set:
    name: ZOAUSER.GDG
    type: gdg
    state: present
```

New type option is `gdg` which given a GDG base name will create a new GDG (Generation Data Group) which is a catalog entry but won't create any GDS unless a GDG relative notation is used.

```python
# Sample of soon to be implemented gdg function
if DataSet.is_gds_relative_name(name):
    src = DataSet.resolve_gds_absolute_name(src)
```

Symbols support
-----------------
 
Changes done to the BetterArgParser should be adopted in the module to not break and use the correct escaped name in case of the data set name having a symbol in it.

Add symbol support to the name option.

Also need to make sure that it works with batch option.

data set name with symbol

`ZOAUSER.TE$T`

That way, the modules will have to retrieve the new escaped named and the original user provided name inside the dictionary like:

```python
name = DataSet.escape_data_set_name(name)
```

Escaping special characters could be done via a new function inside the BetterArgParser.
These are some of the special characters:
- \#
- @
- $
- \-

For this task completion we need:
- [ ] Do the development work
- [ ] Update the docs to talk about GDG and GDS handling in the module.
- [ ] Update the docs with at least 1 example on how to use with GDSs.
- [ ] Create a new test case for GDSs.
- [ ] Create a new test case for symbols.

### Ansible module

_No response_

### PR DESCRIPTION
**Pull request by fernandofloresg**
##### SUMMARY

Add support of GDG and special characters to perform data set operations.

This includes an architectural change to how the module performs the operations.

Adds new tests for data set with special characters.

#### MVSDataSet class
Represents a z/OS Data set, wanted to use DataSet name but is taken.
During initialization, If the data set name is a GDS relative name it will get resolved into a GDS absolute name.

Although it calls other DataSet class static methods, the importance of this class is that it will help simplify how we handle data sets in modules, instead of having many variables around that belong to the data set attributes we can now encapsulate them into a MVSDataSet object.

```python
data_set = MVSDataSet(
            name=params.get("name"),
            record_format=params.get("record_format", None),
            volumes=params.get("volumes", None),
            data_set_type=params.get("type", None),
            block_size=params.get("block_size", None),
            record_length=params.get("record_length", None),
            space_primary=params.get("space_primary", None),
            space_secondary=params.get("space_secondary", None),
            space_type=params.get("space_type", None),
            directory_blocks=params.get("directory_blocks", None),
            key_length=params.get("key_length", None),
            key_offset=params.get("key_offset", None),
            sms_storage_class=params.get("sms_storage_class", None),
            sms_data_class=params.get("sms_data_class", None),
            sms_management_class=params.get("sms_management_class", None),
        )
data_set.ensure_present()

# get attributes to use in another function 
attributes = vars(data_set)

# delete the created data set
data_set.ensure_absent()

```

A future improvement is to turn `name` attribute into an optional parameter to allow just calling MVSDataSet(**attributes) to create a temporary data set. This will allow us to standardize the many functions to create temporary data sets in the modules.

**methods**:
 create
ensure_present
delete
ensure_absent
ensure_catalog
catalog
ensure_uncatalog
uncatalog
set_state

#### Member class

Represents a data set member, at its current state it only provides methods to create and delete, hope this can be expanded when working on `zos_copy`.

**methods**:
ensure_absent
ensure_present

#### GenerationDataGroup class
Represents a Generation Data Group base, provides methods to create and remove gdg bases.
Provides an interface to easily handle GDGs in the ibm_zos_core collection.

**methods**
create
ensure_present
ensure_absent
clear

Fixes #1371 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

zos_data_set

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by fernandofloresg** [comment]
Screenshot 2024-05-14 at 6 03 59 AM
Screenshot 2024-05-14 at 6 04 22 AM

**Review by richp405 — APPROVED** [review]
Looks fine, although there may be another change to helpers/dataset involving how the symbols are loaded.

File: plugins/module_utils/data_set.py [hunk 1]
```diff
@@ -1850,14 +1855,386 @@ def __init__(
...
         self.volumes = volumes
         self.is_gds_active = False
+        self.is_cataloged = False
+
         # If name has escaped chars or is GDS relative name we clean it.
-        self.name = DataSet.escape_data_set_name(self.name)
+        # self.name = DataSet.escape_data_set_name(self.name)
         if DataSet.is_gds_relative_name(self.name):
             try:
                 self.name = DataSet.resolve_gds_absolute_name(self.name)
```
Comment (AndreMarcel99) [inline plugins/module_utils/data_set.py:1864]: Here also to use the `is_gds_active` as a False al the time if resolve the gdg set `is_gds_active` to True

**Review by AndreMarcel99 — CHANGES_REQUESTED** [review]

File: plugins/module_utils/data_set.py (hunk 1 above)
Comment (fernandofloresg) [reply plugins/module_utils/data_set.py:1864]: Nice catch

File: plugins/modules/zos_data_set.py [hunk 2]
```diff
@@ -1036,16 +1152,53 @@ def data_set_type(contents, dependencies):
     #     return None
     if contents is None:
         return "pds"
+    # raise ValueError (
```
Comment (rexemin) [inline plugins/modules/zos_data_set.py:1155]: We can erase this, right?

File: plugins/modules/zos_data_set.py [hunk 3]
```diff
@@ -1036,16 +1152,53 @@ def data_set_type(contents, dependencies):
...
+    #         vars(dependencies)
+    #     )
+    if contents == "gdg" and dependencies.get("state") == "present" and dependencies.get("limit") is None:
+        raise ValueError(
+            "Limit must be provided when data set type is gd and state=present."
```
Comment (rexemin) [inline plugins/modules/zos_data_set.py:1160]: A typo here: `gd`

File: plugins/modules/zos_data_set.py [hunk 4]
```diff
@@ -1036,16 +1152,53 @@ def data_set_type(contents, dependencies):
...
         raise ValueError(
-            "Value {0} is invalid for type argument. type must be of of the following: {1}.".format(
+            "Value {0} is invalid for type argument. type must be one of the following: {1}.".format(
                 contents, ", ".join(DATA_SET_TYPES)
             )
         )
     return contents
 
 
+def limit_type(contents, dependencies):
```
Comment (rexemin) [inline plugins/modules/zos_data_set.py:1172]: Is this used in the `BetterArgParser`? If so, would it be better to move this function into the arg parser util?

**Review by rexemin — CHANGES_REQUESTED** [review]
Looks really good, just have some questions and mostly doc corrections

**Review by AndreMarcel99 — APPROVED** [review]
Looks well

File: plugins/modules/zos_data_set.py (hunk 4 above)
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py:1172]: I don't think it makes sense to be in BetterArgParser, limit option is only used when creating a gdg without any reference. Like in `zos_copy`, lets say we support copying a gdg base to another gdg base, we can check the limit option for the `src`.
Backup and restore, unarchive should do that automatically as well.

File: plugins/modules/zos_data_set.py (hunk 2 above)
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py:1155]: lol that was part of my debugging!!

File: plugins/modules/zos_data_set.py [hunk 5]
```diff
@@ -232,6 +238,54 @@
...
+      - If true removes all gds entries from GDG base when a new gds is created that causes the
+        gdg limit to be exceeded.
+      - Default is false.
+    type: bool
+    required: false
+  extended:
+    description:
+      - Sets the I(extended) attribute for Generation Data Groups.
+      - If false allow up to 255 generation data sets (GDSs) to be associated with the GDG.
+      - If true allow up to 999 generation data sets (gds) to be associated with the gdg.
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py:254]: you are correct, that should be GDSs

File: plugins/module_utils/data_set.py [hunk 6]
```diff
@@ -1850,14 +1855,387 @@ def __init__(
...
+            If the GDG base has active generations, remove them and delete the GDG base.
+
+        Returns
+        -------
+        int
+            Indicates if changes were made.
+        """
+        # Try to delete
+        rc = datasets.delete(self.name)
+        if rc > 0:
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/module_utils/data_set.py:2210]: that would be taken care in line 2218

[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/module_utils/data_set.py:2210]: done

File: plugins/modules/zos_data_set.py [hunk 7]
```diff
@@ -232,6 +238,54 @@
       - I(key_length) should only be provided when I(type=ksds)
     type: int
     required: false
+  empty:
+    description:
+      - Sets the I(empty) attribute for Generation Data Groups.
+      - If false removes only the oldest gds entry when a new gds is created that causes gdg limit to be exceeded.
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py]: done

File: plugins/modules/zos_data_set.py [hunk 8]
```diff
@@ -232,6 +238,54 @@
...
+      - If true removes all gds entries from GDG base when a new gds is created that causes the
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py]: done

File: plugins/modules/zos_data_set.py [hunk 9]
```diff
@@ -232,6 +238,54 @@
...
+        gdg limit to be exceeded.
+      - Default is false.
+    type: bool
+    required: false
+  extended:
+    description:
+      - Sets the I(extended) attribute for Generation Data Groups.
+      - If false allow up to 255 generation data sets (GDSs) to be associated with the GDG.
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py]: done

File: plugins/modules/zos_data_set.py [hunk 10]
```diff
@@ -232,6 +238,54 @@
...
+      - If true allow up to 999 generation data sets (gds) to be associated with the gdg.
+      - Default is false.
+    type: bool
+    required: false
+  fifo:
+    description:
+      - Sets the I(fifo) attribute for Generation Data Groups.
+      - If false the order is the newest gds defined to the oldest gds. this is the default value..
```
[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py]: done

[the remark this answers is not in the document]
Comment (fernandofloresg) [reply plugins/modules/zos_data_set.py]: even here I had added to points at the end

**Review by rexemin — APPROVED** [review]

**Pull request merged**